main.js

```diff
@@ -217,34 +216,108 @@ function updateDevices() {
-    'm3': 'value.volume',
-    'sec': 'value.duration',
-    'hours': 'value.duration',
-    'minutes': 'value.duration'
-};
-
-const UNITS2UNITS = {
-    'deg C': '°C',
-    'deg K': '°K',
-    'deg F': '°F',
-    'seconds': 'sec',
-    'm m^3/h': 'm3/h',
-    'm m^3/min': 'm3/min',
-    'm m^3/s': 'm3/s',
-    'm^3': 'm3',
-    'm m^3': 'm3'
-};
+
+function adjustUnit(unit, type, forcekWh) {
+    let m;
+    // regex depending on type as to account for different units and keep it somewhat readable
+    switch (type) {
+        case "Energy": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(Wh|J)$/); break;
+        case "Mass": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(kg)$/); break;
+        case "Power": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(W|J\/h)$/); break;
+        case "Volume": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( m\^3)$/); break;
+        case "Volume flow": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( m\^3\/(h|min|s))$/); break;
+        case "Mass flow": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( kg\/h)$/); break;
+        case "Flow temperature":
+        case "Return temperature": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?(deg (C|F))$/); break;
+        case "External temperature":
+        case "Temperature Difference": m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( deg (C|F))$/); break;
+        case "Pressure":  m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( bar)$/); break;
+        case "Voltage":  m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( V)$/); break;
+        case "Current":  m = unit.match(/^([0-9e\+\-]+)?\s?([A-Za-z]+)?( A)$/); break;
+        case "Time Point": return {factor: undefined, unit: undefined};
+    }
+    //special case to adjust unit for durations
+    switch (unit) {
+        case "seconds": return {factor: 1, unit: "s"};
+        case "minutes": return {factor: 1, unit: "min"};
+        case "hours": return {factor: 1, unit: "h"};
+        case "days": return {factor: 1, unit: "d"};
+    }
+
+    // nothing worked
+    if (!m) {
+        return {factor: undefined, unit: undefined};
+    }
+
+    // adjust factor depending on metric prefix
+    let factor = parseFloat(m[1]) || 1;
+    unit = m[3].trim();
+    if (m[2]) {
+        switch (m[2]) {
+            case "m": factor *= 1e-3; break;
+            case "my": factor *= 1e-6; break;
+            case "k": factor *= 1e3; break;
+            case "M": factor *= 1e6; break;
+            case "G": factor *= 1e9; break;
+            case "T": factor *= 1e9; break; //this is an error in libmbus...
+            default: unit = m[2] + unit;
+        }
+    }
+
+    // make some units nicer looking
+    switch (unit) {
+        case "deg C": unit = "°C"; break;
+        case "deg F": unit = "°F"; break;
+        case "m^3": unit = "m³"; break;
+        case "m^3/h": unit = "m³/h"; break;
+        case "m^3/min": unit = "m³/min"; break;
+        case "m^3/s": unit = "m³/s"; break;
+    }
+    switch (type) {
```

Comment:
You expect more such special cases? If not a traditional if construict is better readable, or? :-)

Reply:
True ;) I started the code that way because I expected more cases, but found none

Reply:
:-))